// @bundle-stats/utils/lib/lighthouse/metrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.metrics = void 0;
const constants_1 = require("../constants");
exports.metrics = {
    score: {
        label: 'Score',
        type: constants_1.MetricTypes.Score,
    },
    performanceScore: {
        label: 'Performance',
        type: constants_1.MetricTypes.Score,
    },
    accessibilityScore: {
        label: 'Accessibility',
        type: constants_1.MetricTypes.Score,
    },
    bestPracticesScore: {
        label: 'Best Practices',
        type: constants_1.MetricTypes.Score,
    },
    seoScore: {
        label: 'SEO',
        type: constants_1.MetricTypes.Score,
    },
    pwaScore: {
        label: 'PWA',
        type: constants_1.MetricTypes.Score,
    },
    speedIndex: {
        label: 'Perceptual Speed Index',
        type: constants_1.MetricTypes.Duration,
    },
    firstMeaningfulPaint: {
        label: 'First Meaningful Paint',
        type: constants_1.MetricTypes.Duration,
    },
    timeToFirstByte: {
        label: 'Time To First Byte',
        type: constants_1.MetricTypes.Duration,
    },
    firstInteractive: {
        label: 'First Interactive',
        type: constants_1.MetricTypes.Duration,
    },
    totalByteWeight: {
        label: 'Total Weight',
        type: constants_1.MetricTypes.FileSize,
    },
    domSize: {
        label: 'DOM Size',
        type: constants_1.MetricTypes.Numeric,
    },
};
