// @bundle-stats/utils/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JobSectionId = exports.DeltaType = exports.MetricTypes = exports.InsightType = exports.Source = void 0;
var Source;
(function (Source) {
    Source["webpack"] = "webpack";
    Source["lighthouse"] = "lighthouse";
    Source["browsertime"] = "browsertime";
})(Source || (exports.Source = Source = {}));
var InsightType;
(function (InsightType) {
    InsightType["ERROR"] = "error";
    InsightType["INFO"] = "info";
    InsightType["WARNING"] = "warning";
})(InsightType || (exports.InsightType = InsightType = {}));
var MetricTypes;
(function (MetricTypes) {
    MetricTypes["Numeric"] = "METRIC_TYPE_NUMERIC";
    MetricTypes["Score"] = "METRIC_TYPE_SCORE";
    MetricTypes["FileSize"] = "METRIC_TYPE_FILE_SIZE";
    MetricTypes["Duration"] = "METRIC_TYPE_DURATION";
    MetricTypes["Percentage"] = "METRIC_TYPE_PERCENTAGE";
})(MetricTypes || (exports.MetricTypes = MetricTypes = {}));
var DeltaType;
(function (DeltaType) {
    DeltaType["HIGH_NEGATIVE"] = "HIGH_NEGATIVE";
    DeltaType["NEGATIVE"] = "NEGATIVE";
    DeltaType["LOW_NEGATIVE"] = "LOW_NEGATIVE";
    DeltaType["NO_CHANGE"] = "NO_CHANGE";
    DeltaType["LOW_POSITIVE"] = "LOW_POSITIVE";
    DeltaType["POSITIVE"] = "POSITIVE";
    DeltaType["HIGH_POSITIVE"] = "HIGH_POSITIVE";
})(DeltaType || (exports.DeltaType = DeltaType = {}));
var JobSectionId;
(function (JobSectionId) {
    JobSectionId["meta"] = "meta";
    JobSectionId["insights"] = "insights";
    JobSectionId["summary"] = "summary";
    JobSectionId["metrics"] = "metrics";
    JobSectionId["rawData"] = "rawData";
})(JobSectionId || (exports.JobSectionId = JobSectionId = {}));
